File: Python/challenge_3.py
"""
Create simple class that converts roman numerals to integers and vice versa.

"IV" = 4
9 = "IX"
etc.

Refer to the unit tests to grasp the idea of this task.
Feel free to add as many unit tests as you want.
No external dependencies!

P.S.: you cant leverage all benefits of classes. ;) 
"""
import unittest


class Convertor:
    """
    Base class that converts roman numerals to integers and vice versa.
    """

    @staticmethod
    def convert_to_roman(integer: int) -> str:
        """
        Integers to roman numerals
        """
        """
                Roman numerals to integers
                """
        # The arabic number to be converted
        t = int(integer)
        # The list contains the number's digits
        l = list()

        # int is not iterable, so I wrote this to get the digits
        while t >= 1:
            l.append(int(t % 10))
            t = t / 10

        # They are in the wrong order, so the list has to be reversed
        l.reverse()

        # The format
        format = ""

        # Current position in the number
        c = 0

        # If the number is at least 1000
        if len(l) >= 4:
            num = 0

        # All the digits after 1000
        i = int(len(l) - 1)
        while i > 3:
            num = num + 1
            i = i - 1

            # A list containing these digits
        li = list()

        for i in range(0, num + 1):
            li.append(l[i])

        # Converts the list to an int
        n = map(str, li)
        n = ''.join(n)
        n = int(n)

        # Add an 'M' for every 1000
        for i in range(0, n):
            format = format + 'M'

        c = num + 1

        # If the number is also least 100
        if len(l) >= 3:
            format = format + encode(l, c, 'C', 'D', "CM")
            c = c + 1

        # If the number is at least 10
        if len(l) >= 3:
            format = format + encode(l, c, 'X', 'L', "XC")
            c = c + 1

        # If the number is at least 100 but smaller than 1000
        if len(l) > 2:
            format = format + encode(l, c, 'I', 'V', "IX")

        # If the number is at least 10 but smaller than 100
        if len(l) == 2:
            format = format + encode(l, c, 'X', 'L', "XC")
            c = c + 1
            format = format + encode(l, c, 'I', 'V', "IX")

        # If the number is at least 1 but smaller than 10
        if len(l) == 1:
            format = format + encode(l, c, 'I', 'V', "IX")

        return format
        return ""

    @staticmethod
    def convert_to_integers(roman):
        # Hold the numbers in the list
        l = list()

        # The format
        format = 0

        # Convert from roman numeral to arabic number
        for i in range(0, len(roman)):
            if roman[i] == 'I':
                l.append(int(1))
            elif roman[i] == 'V':
                l.append(int(5))
            elif roman[i] == 'X':
                l.append(int(10))
            elif roman[i] == 'L':
                l.append(int(50))
            elif roman[i] == 'C':
                l.append(int(100))
            elif roman[i] == 'D':
                l.append(int(500))
            elif roman[i] == 'M':
                l.append(int(1000))

        # Calculate the format as follows:
        i = 0
        while i < len(l) - 1:
            # If the next numeral is greater than the current one, add the next one minus the current one = format = format + next_numeral - current_numeral
            if l[i] < l[i + 1]:
                format = format + l[i + 1] - l[i];
                i = i + 2
            # Else add it normally
            else:
                format = format + l[i]
                i = i + 1

        # Fix in case the last two numerals are equal
        if i == len(l):
            if l[i - 2] == l[i - 1]:
                format = format + l[i - 1]
        else:
            format = format + l[i]

        # Return the format
        return format


def encode(l, c, one, five, nine):
    # The format to be returned
    format = ""

    # Special case 5.1: 5 has its own numeral 'V'
    # Special case 5.2: 50 has its own numeral 'A'
    # Special case 5.3: 500 has its own numeral 'D'
    if l[c] == 5:
        format = format + five
    elif l[c] < 5:
        for i in range(0, l[c]):
            format = format + one
    elif l[c] > 5:
        # Special case 9.1: 9 depends on 10, so it's 'IX'
        # Special case 9.2: 90 depends on 1000, so it's 'XC'
        # Special case 9.3: 900 depends on 1000, so it's 'CM'
        if l[c] == 9:
            format = format + nine
        else:
            format = format + five
            for i in range(6, l[c] + 1):
                format = format + one

    return format


class ConvertorTestCase(unittest.TestCase):
    """
    Unit tests
    """

    def __init__(self, methodName: str = "runTest") -> None:
        """
        Superconstructor
        """
        self.convertor = Convertor()
        self.test_cases = [
            {
                "integer": 1995,
                "roman": "MCMXCV",
            },
            {
                "integer": 2023,
                "roman": "MMXXIII",
            },
        ]
        super().__init__(methodName)

    def test_convert_to_roman(self):
        """
        Table-driven tests
        """
        for case in self.test_cases:
            res = self.convertor.convert_to_roman(case["integer"])
            self.assertEqual(
                res,
                case["roman"],
                f'\nfailed test case "{case["integer"]}": expected "{case["roman"]}", but got "{res}".',
            )

    def test_convert_to_integers(self):
        """
        Table-driven tests
        """
        for case in self.test_cases:
            res = self.convertor.convert_to_integers(case["roman"])
            self.assertEqual(
                res,
                case["integer"],
                f'\nfailed test case "{case["roman"]}": expected "{case["integer"]}", but got "{res}".',
            )


if __name__ == "__main__":
    unittest.main()

File: C#/CodeFights/CodeFightsTests/Challenge_3.cs
/*
 Create simple class that converts roman numerals to integers and vice vers

"IV" = 4
9 = "IX"
etc.

Refer to the unit tests to grasp the idea of this task.
Feel free to add as many unit tests as you want.
No external dependencies!

P.S.: you cant leverage all benefits of classes. ;)
*/

using System.Linq;
using System.Text;

namespace CodeFights
{
    using NUnit.Framework;
    using System;

    [TestFixture]
    public class Challenge_3
    {
        [Test]
        public void TestToRoman()
        {
            Assert.AreEqual("I", ConvertToRoman(1), "1 should convert to 'I'");
            Assert.AreEqual("II", ConvertToRoman(2), "2 should convert to 'II'");
            Assert.AreEqual("IV", ConvertToRoman(4), "4 should convert to 'IV'");
            Assert.AreEqual("V", ConvertToRoman(5), "5 should convert to 'V'");
            Assert.AreEqual("VI", ConvertToRoman(6), "6 should convert to 'VI'");
            Assert.AreEqual("IX", ConvertToRoman(9), "9 should convert to 'IX'");
            Assert.AreEqual("X", ConvertToRoman(10), "10 should convert to 'X'");
            Assert.AreEqual("XXIII", ConvertToRoman(23), "23 should convert to 'XXIII'");
            Assert.AreEqual("XLIV", ConvertToRoman(44), "44 should convert to 'XLIV'");
            Assert.AreEqual("LIX", ConvertToRoman(59), "59 should convert to 'LIX'");
            Assert.AreEqual("LXXV", ConvertToRoman(75), "75 should convert to 'LXXV'");
            Assert.AreEqual("XCII", ConvertToRoman(92), "92 should convert to 'XCII'");
            Assert.AreEqual("CCLXI", ConvertToRoman(261), "261 should convert to 'CCLXI'");
            Assert.AreEqual("M", ConvertToRoman(1000), "1000 should convert to 'M'");
            Assert.AreEqual("MCDL", ConvertToRoman(1450), "1450 should convert to 'MCDL'");
            Assert.AreEqual("MCMXC", ConvertToRoman(1990), "1990 should convert to 'MCMXC'");
            Assert.AreEqual("MMVIII", ConvertToRoman(2008), "2008 should convert to 'MMVIII'");
            Assert.AreEqual("MMDXXXVII", ConvertToRoman(2537), "2537 should convert to 'MMDXXXVII'");
            Assert.AreEqual("MMMCMXCIX", ConvertToRoman(3999), "3999 should convert to 'MMMCMXCIX'");
        }

        [Test]
        public void TestFromRoman()
        {
            Assert.AreEqual(1, ConvertFromRoman("I"), "'I' should convert to 1");
            Assert.AreEqual(2, ConvertFromRoman("II"), "'II' should convert to 2");
            Assert.AreEqual(4, ConvertFromRoman("IV"), "'IV' should convert to 4");
            Assert.AreEqual(5, ConvertFromRoman("V"), "'V' should convert to 5");
            Assert.AreEqual(6, ConvertFromRoman("VI"), "'VI' should convert to 6");
            Assert.AreEqual(9, ConvertFromRoman("IX"), "'IX' should convert to 9");
            Assert.AreEqual(10, ConvertFromRoman("X"), "'X' should convert to 10");
            Assert.AreEqual(23, ConvertFromRoman("XXIII"), "'XXIII' should convert to 23");
            Assert.AreEqual(44, ConvertFromRoman("XLIV"), "'XLIV' should convert to 44");
            Assert.AreEqual(59, ConvertFromRoman("LIX"), "'LIX' should convert to 59");
            Assert.AreEqual(75, ConvertFromRoman("LXXV"), "'LXXV' should convert to 75");
            Assert.AreEqual(92, ConvertFromRoman("XCII"), "'XCII' should convert to 92");
            Assert.AreEqual(261, ConvertFromRoman("CCLXI"), "'CCLXI' should convert to 261");
            Assert.AreEqual(1000, ConvertFromRoman("M"), "'M' should convert to 1000");
            Assert.AreEqual(1450, ConvertFromRoman("MCDL"), "'MCDL' should convert to 1450");
            Assert.AreEqual(1990, ConvertFromRoman("MCMXC"), "'MCMXC' should convert to 1990");
            Assert.AreEqual(2008, ConvertFromRoman("MMVIII"), "'MMVIII' should convert to 2008");
            Assert.AreEqual(2537, ConvertFromRoman("MMDXXXVII"), "'MMDXXXVII' should convert to 2537");
            Assert.AreEqual(3999, ConvertFromRoman("MMMCMXCIX"), "'MMMCMXCIX' should convert to 3999");
        }

        private static string _placeholderToRoman = "{0} should convert to {1}";
        public static string ConvertToRoman(int n)
        {
            var result = new StringBuilder();
            Enumerable.Range(0, n / 1000).Select(x => x).ToList().ForEach(x => result.Append("M"));
            n = n % 1000;
            n = HandleCase(n, result, 900, "CM");
            n = HandleCase(n, result, 500, "D");
            n = HandleCase(n, result, 400, "CD");
            Enumerable.Range(0, n / 100).Select(x => x).ToList().ForEach(x => result.Append("C"));
            n = n % 100;
            n = HandleCase(n, result, 90, "XC");
            n = HandleCase(n, result, 50, "L");
            n = HandleCase(n, result, 40, "XL");
            Enumerable.Range(0, n / 10).Select(x => x).ToList().ForEach(x => result.Append("X"));
            n = n % 10;
            n = HandleCase(n, result, 9, "IX");
            n = HandleCase(n, result, 5, "V");
            n = HandleCase(n, result, 4, "IV");
            Enumerable.Range(0, n / 1).Select(x => x).ToList().ForEach(x => result.Append("I"));

            return result.ToString();
        }

        private static int HandleCase(int n, StringBuilder result, int limit, string input)
        {
            if (n >= limit)
            {
                result.Append(input);
                n = n - limit;
            }
            return n;
        }

        public static int ConvertFromRoman(string romanNumeral)
        {
            var result = romanNumeral
                .Replace("CM", "A")
                .Replace("CD", "B")
                .Replace("XC", "W")
                .Replace("XL", "Q")
                .Replace("IX", "Y")
                .Replace("IV", "Z");

            int sum = 0;
            foreach (var character in result)
            {
                sum += GetFromChar(character);
            }

            return sum;
        }

        private static int GetFromChar(char c)
        {
            return c switch
            {
                'M' => 1000,
                'D' => 500,
                'C' => 100,
                'L' => 50,
                'X' => 10,
                'V' => 5,
                'I' => 1,
                'A' => 900,
                'B' => 400,
                'W' => 90,
                'Q' => 40,
                'Y' => 9,
                'Z' => 4,
            };
        }
    }
}